### PR TITLE
feat(storage): support soft delete timestamps

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #13575

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13728)

### CODE REVIEW

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/object_metadata.h [hunk 1]
```diff
@@ -505,6 +505,50 @@ class ObjectMetadata {
     return *this;
   }
 
+  /// Returns true if the object has a soft delete timestamp.
+  bool has_soft_delete_time() const { return soft_delete_time_.has_value(); }
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_metadata.h:509]: All other timestamps (and all other scalars really) return a default value when not set.  If we had a time machine I would probably go back and do something different, but I am not sure it is worth changing the approach on only new fields.

File: google/cloud/storage/object_metadata.h [hunk 2]
```diff
@@ -505,6 +505,50 @@ class ObjectMetadata {
...
+  }
+
+  /// @note This is only intended for mocking.
+  ObjectMetadata& set_hard_delete_time(
+      std::chrono::system_clock::time_point v) {
+    hard_delete_time_ = v;
+    return *this;
+  }
+
+  /// @note THis is only intended for mocking.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_metadata.h]: Fixed.

File: google/cloud/storage/object_metadata.h [hunk 3]
```diff
@@ -505,6 +505,50 @@ class ObjectMetadata {
...
+    return *this;
+  }
+
+  /// @note This is only intended for mocking.
+  ObjectMetadata& reset_soft_delete_time() {
+    soft_delete_time_.reset();
+    return *this;
+  }
+
+  /// Returns true if the object has a soft delete timestamp.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_metadata.h]: Fixed.

File: google/cloud/storage/object_metadata.h [hunk 4]
```diff
@@ -505,6 +505,50 @@ class ObjectMetadata {
...
+  bool has_hard_delete_time() const { return hard_delete_time_.has_value(); }
+
+  /// The timestamp when the object will be hard deleted. Objects that are not
+  /// soft deleted return a default initialized time point.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_metadata.h]: Fixed I hope.

**Pull request merged**